### DISCUSSION TITLE
Multi language support

### DISCUSSION THREAD

**Discussion by Manubi** [nextauthjs/next-auth#1129] [Help] [answered]
Hey,
do you plan on supporting multiple languages and custom login forms/emails for the auth flow? If so, I’d be happy to help.

best 
Manuel

**Comment by balazsorban44** [comment, marked as answer]
Hi there, we already support custom pages (including login forms, only it is called signing):

https://next-auth.js.org/configuration/pages

and you can override the email template with the `sendVerificationRequest` method:
https://next-auth.js.org/providers/email#customising-emails


Here is the built-in signing page for inspiration:
https://github.com/nextauthjs/next-auth/blob/canary/src/server/pages/signin.js

If you need multi-language support, I think you could open a feature request issue, maybe with a proposal on how would you prefer doing it.

**Reply by Manubi** [reply]
Awesome! Thanks!!

**Reply by iaincollins** [reply]
Thanks for asking! FWIW I would love us have better multi-language support, it felt like supporting custom pages was an escape hatch to allow folks to support themselves, until we can get to it.

With [built-in i18n routing now in in Next.js](https://nextjs.org/docs/advanced-features/i18n-routing) perhaps we are in a better position to support it out of the box.

I guess we'd leverage the `NEXT_LOCALE` cookie if set and otherwise fall back to `Accept-Language` header to get the preferred locale of the client. When we come to multiple domain support, we could could also look for and parse next.config.js to get the default locale for a given domain.

**Reply by stevedaniels-traintime** [reply]
We are currently working on a project and have implemented NextAuth ✨ 

However we also require internationalisation via i18n as we are launching a multi-lingual platform.

It would be great if anyone has any ideas how we could implement it so that we can continue to use NextAuth 🙏

**Reply by iaincollins** [reply]
If you use custom sign in pages there are no blockers for internationalisation / localisation as all pages can be customised.

The only caveat to this might be if you are using the email provider; while you can customise the email message the locale isn't explicitly passed to the to the `sendVerificationRequest` so you'd have get a bit clever there and add logic into `[...nextauth].js]` to handle sending a different email body depending on what the request headers / cookies look like.

If anyone would like any tips on that I'm happy to elaborate.

**Reply by YuriGor** [reply]
I didn't think about emails yet, but there isone more point with custom pages vs Nextjs10 localized routes.

Next-auth `signin` method does not respect all this /en/ /fr/ prefixes and always opens custom login page directly by URL as configured in [...nextauth].js

As a workaround I don't use `signin` method, but instead I use next/router.
Here is an example:
https://github.com/nextauthjs/next-auth/issues/1154#issuecomment-782132611

For the rest part of i18n story next-auth works perfectly.

**Reply by darshkpatel** [reply]
hey @iaincollins I'd love to know more about passing language in the sendVerificationRequest call, care to elaborate ?

**Reply by gamessty** [reply]
In order to customize the emails of the user, I managed to get the locale using the callback url passed to the `sendVerificationRequest ` as it contained the language the user was on in the first path. Perhaps a cookie/ request headers may be a more "by the book" solution, but in my case this workaround works very well for most cases. I am also curious if anyone might implemen it.
